src/crypto/tls/delegated_credentials_test.go

```diff
@@ -0,0 +1,413 @@
+
+	// Test validation of Delegated Credential which TTL is short
+	delegatedCred, _, err = NewDelegatedCredential(cert, ECDSAWithP256AndSHA256, shortValidTime, "server")
+	if err != nil {
+		t.Fatal(err)
+	}
+	if !delegatedCred.Validate(cert.Leaf, "server", dcNow) {
+		t.Error("valid Delegated Credential is invalid; want valid")
+	}
+
+	//cert.Leaf.SignatureAlgorithm = x509.ECDSAWithSHA256
+	//delegatedCred.Algorithm = ECDSAWithP521AndSHA512
+
+	// Test signature algorithm binding
+	//if delegatedCred.Validate(cert.Leaf, "server", dcNow) {
+	//	t.Error("Delegated Credential with wrong scheme is valid; want invalid")
+	//}
+
+	delegatedCred.Algorithm = ECDSAWithP256AndSHA256
+
+	// Test delegation certificate binding
+	cert.Leaf.Raw[0] ^= byte(42)
+	if delegatedCred.Validate(cert.Leaf, "server", dcNow) {
+		t.Error("Delegated Credential with wrong certificate is valid; want invalid")
+	}
+
+	// Test validation of DC using a certificate that can't delegate.
+	if delegatedCred.Validate(dcTestCerts["no dc"].Leaf, "server", dcNow) {
+		t.Error("Delegated Credential with non-delegation cert is valid; want invalid")
+	}
+}
+
+// Test encoding/decoding of delegated credentials.
+func TestDelegatedCredentialMarshal(t *testing.T) {
+	initialize()
+	cert := dcTestCerts["dc"]
+	time := dcNow.Sub(cert.Leaf.NotBefore) + dcMaxTTL
+
+	delegatedCred, _, err := NewDelegatedCredential(cert, ECDSAWithP256AndSHA256, time, "server")
+	if err != nil {
+		t.Fatal(err)
+	}
+
+	ser, err := delegatedCred.Marshal()
+	if err != nil {
+		t.Error(err)
+	}
+
+	delegatedCred2, err := UnmarshalDelegatedCredential(ser)
+	if err != nil {
+		t.Error(err)
+	}
+
+	err = equal(delegatedCred, delegatedCred2)
+	if err != nil {
+		t.Error(err)
+	}
+
+	if delegatedCred.Algorithm != delegatedCred2.Algorithm {
+		t.Errorf("scheme mismatch: got %04x; want %04x", delegatedCred2.Algorithm, delegatedCred.Algorithm)
+	}
+
+	if !bytes.Equal(delegatedCred2.Signature, delegatedCred.Signature) {
+		t.Error("Signature mismatch")
+	}
+}
+
+var dcTests = []struct {
+	clientDCSupport  bool
+	serverDCSupport  bool
+	clientSkipVerify bool
+	clientMaxVers    uint16
+	serverMaxVers    uint16
+	expectSuccess    bool
+	expectDC         bool
+	name             string
+}{
+	{true, true, false, VersionTLS13, VersionTLS13, true, true, "tls13"},
+	{true, true, true, VersionTLS13, VersionTLS13, true, true, "tls13, using server skip verify"},
+	{false, true, false, VersionTLS13, VersionTLS13, true, false, "client no DC"},
+	{true, false, false, VersionTLS13, VersionTLS13, true, false, "server no DC"},
+	{true, false, false, VersionTLS12, VersionTLS13, true, false, "client using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS13, VersionTLS12, true, false, "server using TLS 1.2. No DC is supported in that version."},
+	{true, false, false, VersionTLS11, VersionTLS13, true, false, "client using TLS 1.1. No DC is supported in that version."},
```

Comment:
Perhaps this should be TLS 1.2 instead of TLS 1.1?

Reply:
There is already a test for tls1.2 on line 284... or what do you mean?

Reply:
Well, I could imagine a bug wherein we disallow DCs with client=1.3 and server=1.2, but not the other way around. I'm suggesting we test both (client=1.3, server=1.2) and (client=1.2, server=1.3). Testing 1.2 will supersedes TLS 1.1.

Reply:
Mmm, there are already two tests:

On line 284: Client is 1.12 and Server is 1.13
On line 285: Client is 1.13 and Server is 1.12

And there is also on line 286:
Client is 1.1 and Server is 1.13

Reply:
Oh, thanks! I didn't see the one on 284.